File: Dia12/d12p1/src/main.py
def path(node, connections, visited):
	if node == "end":
		return [["end"]]
	res = []
	visited = visited.copy()
	visited.append(node)
	for c in connections[node]:
		if c.isupper() or c not in visited:
			# print(node, ">", c)
			for x in path(c, connections, visited):
				x.insert(0, node)
				res.append(x)
	return res



with open('../../input.txt') as f:
	connections = {}
	for [n0, n1] in map(lambda x: x.strip().split('-'), f.readlines()):
		if n0 not in connections:
			connections[n0] = []
		if n1 not in connections:
			connections[n1] = []
		connections[n0].append(n1)
		connections[n1].append(n0)
	# print(connections["start"])
	print(len(path("start", connections, [])))

File: Dia12/d12p1/src/main.rs
use std::{collections::HashMap};

fn possible_paths<'a>(
    node: &'a str,
    connections: &'a HashMap<&str, Vec<&str>>,
    mut visited: Vec<&'a str>,
) -> Vec<Vec<&'a str>> {
    if node == "end" {
        return vec![vec!["end"]];
    }
    visited.push(node);
    connections[node]
        .iter()
        .filter(|x| !visited.contains(x) || x.to_uppercase() == x.to_string())
        .flat_map(|x| {
            // println!("{} {:?}", x, visited);
            possible_paths(*x, connections, visited.clone())
                .into_iter()
                .map(|mut x| {
                    x.insert(0, node);
                    // println!(">>> {:?}", x);
                    x
                })
        }).collect()
}

fn main() {
    let mut connections = HashMap::new();

    for line in include_str!("../../input.txt")
        .lines()
        .map(|x| x.split('-').collect::<Vec<&str>>())
    {
        connections
            .entry(line[0])
            .or_insert_with(Vec::new)
            .push(line[1]);
        connections
            .entry(line[1])
            .or_insert_with(Vec::new)
            .push(line[0]);
    }

    let p = possible_paths("start", &connections, Vec::new());
    println!(">>> {}", p.len())
}
